tls/message.py

```diff
@@ -205,6 +207,73 @@ def from_bytes(cls, bytes):
 
 
 @attr.s
+class URLAndHash(object):
+    """
+    An object representing a URLAndHash struct.
+    """
+    url = attr.ib()
+    padding = attr.ib()
+    sha1_hash = attr.ib()
+
+    # XXX: The "padding" byte MUST be 0x01.  It is
```

Comment:
You could use an [attrs validator](http://attrs.readthedocs.io/en/stable/examples.html#validators) here, as it'd cover both serialization and deserialization.

But since this is a parsing issue, it might be nice to fail early when reading potentially malicious data.  You can use [`construct.OneOf`](http://construct.readthedocs.io/en/latest/adapters.html#construct.OneOf) to do that:


```python
>>> construct.OneOf(construct.UBInt8('a'), [1]).parse('\x01')
1
>>> construct.OneOf(construct.UBInt8('a'), [1]).build(1)
'\x01'
>>> construct.OneOf(construct.UBInt8('a'), [1]).parse('\x02')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/mrw/.virtualenvs/tls/local/lib/python2.7/site-packages/construct/core.py", line 188, in parse
    return self.parse_stream(BytesIO(data))
  File "/home/mrw/.virtualenvs/tls/local/lib/python2.7/site-packages/construct/core.py", line 198, in parse_stream
    return self._parse(stream, Container())
  File "/home/mrw/.virtualenvs/tls/local/lib/python2.7/site-packages/construct/core.py", line 288, in _parse
    return self._decode(self.subcon._parse(stream, context), context)
  File "/home/mrw/.virtualenvs/tls/local/lib/python2.7/site-packages/construct/adapters.py", line 412, in _decode
    raise ValidationError("invalid object", obj)
construct.adapters.ValidationError: ('invalid object', 2)
>>> construct.OneOf(construct.UBInt8('a'), [1]).build(2)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/mrw/.virtualenvs/tls/local/lib/python2.7/site-packages/construct/core.py", line 212, in build
    self.build_stream(obj, stream)
  File "/home/mrw/.virtualenvs/tls/local/lib/python2.7/site-packages/construct/core.py", line 219, in build_stream
    self._build(obj, stream, Container())
  File "/home/mrw/.virtualenvs/tls/local/lib/python2.7/site-packages/construct/core.py", line 290, in _build
    self.subcon._build(self._encode(obj, context), stream, context)
  File "/home/mrw/.virtualenvs/tls/local/lib/python2.7/site-packages/construct/adapters.py", line 415, in _encode
    return self._decode(obj, context)
  File "/home/mrw/.virtualenvs/tls/local/lib/python2.7/site-packages/construct/adapters.py", line 412, in _decode
    raise ValidationError("invalid object", obj)
construct.adapters.ValidationError: ('invalid object', 2)
```

Note that failing early on parsing means failing late on serialization.  Maybe it's not a bad idea to have an attrs validator *and* `construct.OneOf`?